Sources/Jobs/Scheduler/RecurrenceRule.swift

```diff
@@ -0,0 +1,129 @@
+import Foundation
+
+enum RecurrenceRuleError: Error {
+    case atLeastOneRecurrenceRuleConstraintRequiredToIntialize
+    case lowerBoundGreaterThanUpperBound
+    case noSetConstraintForRecurrenceRuleTimeUnit
+    case couldNotResolveDateComponentValueFromRecurrenceRuleTimeUnit
+    case noConstraintsSetForRecurrenceRule
+    case coundNotResolveNextInstanceWithin1000Years
+    case couldNotResolveYearConstraitFromDate
+    case couldNotResloveNextValueFromConstraint
+    case ruleInsatiable
+    case couldNotParseHourAndMinuteFromString
+    case startHourAndMinuteGreaterThanEndHourAndMinute
+}
+
+/// Defines the rule for when to run a job based on the given constraints
+///
+/// - warning: RecurrenceRule only supports the Gregorian calendar (i.e. Calendar.identifier.gregorian or Calendar.identifier.iso8601)
+///
+/// - Note: RecurrenceRule uses the local TimeZone as default
+public struct RecurrenceRule {
+    public enum TimeUnit: CaseIterable {
+        case second
+        case minute
+        case hour
+        case dayOfWeek
+        case dayOfMonth
+        case month
+        case quarter
+        case year
+    }
+
+    var timeZone: TimeZone
+
+    private(set) var yearConstraint: YearRecurrenceRuleConstraint?
+    private(set) var quarterConstraint: QuarterRecurrenceRuleConstraint?
+    private(set) var monthConstraint: MonthRecurrenceRuleConstraint?
+    private(set) var dayOfMonthConstraint: DayOfMonthRecurrenceRuleConstraint?
+    private(set) var dayOfWeekConstraint: DayOfWeekRecurrenceRuleConstraint?
+    private(set) var hourConstraint: HourRecurrenceRuleConstraint?
+    private(set) var minuteConstraint: MinuteRecurrenceRuleConstraint?
+    private(set) var secondConstraint: SecondRecurrenceRuleConstraint?
+
+    private let timeUnitOrder: [RecurrenceRule.TimeUnit] = [
+        .year,
+        .quarter,
+        .month,
+        .dayOfMonth,
+        .dayOfWeek,
+        .hour,
+        .minute,
+        .second
+    ]
+
+    init(timeZone: TimeZone = TimeZone.current) {
```

Comment:
We should find some way to surface this option (via `usingTimeZone` below) in the `ScheduleBuilder` so that users can choose their timezone if they want.